packages/flutter/lib/src/material/material.dart

```diff
@@ -50,33 +125,376 @@ class Material extends StatelessComponent {
-        style: textStyle ?? Theme.of(context).text.body1,
+        style: config.textStyle ?? Theme.of(context).text.body1,
         child: contents
       );
-      if (_kEdges[type] != null) {
-        contents = new ClipRRect(
-          xRadius: _kEdges[type],
-          yRadius: _kEdges[type],
-          child: contents
-        );
-      }
     }
-    return new DefaultTextStyle(
-      style: Theme.of(context).text.body1,
-      child: new AnimatedContainer(
+    contents = new _InkFeatures(
+      key: _inkFeatureRenderer,
+      onPaintNeeded: config.painter != null ? () { setState(() { /* the state that changed is in the renderer */ }); } : null,
+      child: contents
+    );
+    if (config.type == MaterialType.circle) {
+      contents = new ClipOval(child: contents);
+    } else if (kMaterialEdges[config.type] != null) {
+      contents = new ClipRRect(
+        xRadius: kMaterialEdges[config.type],
+        yRadius: kMaterialEdges[config.type],
+        child: contents
+      );
+    }
+    if (config.painter != null) {
+      contents = new CustomPaint(
+        painter: new _InkFeaturesPainter(this, config.painter),
+        child: contents
+      );
+    } else {
+      contents = new AnimatedContainer(
         curve: Curves.ease,
         duration: kThemeChangeDuration,
         decoration: new BoxDecoration(
           backgroundColor: _getBackgroundColor(context),
-          borderRadius: _kEdges[type],
-          boxShadow: elevation == 0 ? null : elevationToShadow[elevation],
-          shape: type == MaterialType.circle ? Shape.circle : Shape.rectangle
+          borderRadius: kMaterialEdges[config.type],
+          boxShadow: config.elevation == 0 ? null : elevationToShadow[config.elevation],
+          shape: config.type == MaterialType.circle ? Shape.circle : Shape.rectangle
         ),
         child: contents
-      )
+      );
+    }
+    return contents;
+  }
+}
+
+class _InkFeaturesPainter extends CustomPainter {
+  const _InkFeaturesPainter(this.materialState, this.painter);
+  final _MaterialState materialState;
+  final CustomMaterialPainter painter;
+  void paint(PaintingCanvas canvas, Size size) {
+    _RenderInkFeatures renderer = materialState._inkFeatureRenderer.currentContext.findRenderObject();
+    painter.paint(canvas, size, renderer.inkFeatures);
+    renderer._paintNeeded = false;
+  }
+  bool shouldRepaint(_InkFeaturesPainter oldDelegate) {
+    assert(materialState == oldDelegate.materialState);
+    assert(materialState?._inkFeatureRenderer?.currentContext?.findRenderObject() != null);
+    _RenderInkFeatures renderer = materialState._inkFeatureRenderer.currentContext.findRenderObject();
+    return renderer._paintNeeded || (painter != oldDelegate.painter && painter.shouldRepaint(oldDelegate.painter));
+  }
+}
+
+const Duration _kHighlightFadeDuration = const Duration(milliseconds: 100);
+
+const double _kDefaultSplashRadius = 35.0 / 2.0;
+const int _kSplashInitialOpacity = 0x30; // 0..255
```

Comment:
We usually use the term "alpha" for the 0..255 units and "opacity" for the 0.0..1.0 units.